community/logging/src/main/java/org/neo4j/logging/RotatingFileOutputStreamSupplier.java

```diff
@@ -190,40 +231,58 @@ private void rotate()
             return;
         }
 
-        Runnable runnable = () -> {
-            OutputStream newStream;
-            try
+        Runnable runnable = () ->
+        {
+            synchronized ( outRef )
```

Comment:
`outRef` can be read-locked for everything except rotation and closing to make sure all those operations can run concurrently. It should be write-locked for rotation and close to ensure that we only rotate or close at a point in time. It does not really matter what we do with the file, some operations can just run concurrently and some can't.